horovod/common/ops/cuda_operations.cc

```diff
@@ -0,0 +1,274 @@
+// limitations under the License.
+// =============================================================================
+
+#include "cuda_operations.h"
+
+#include <thread>
+
+namespace horovod {
+namespace common {
+
+// This event management code is only used with CUDA
+cudaError_t CUDAContext::GetCudaEvent(cudaEvent_t* event) {
+  int device;
+  auto status = cudaGetDevice(&device);
+  if (status != cudaSuccess) {
+    return status;
+  }
+
+  auto& mutex = cuda_events_mutex;
+  {
+    std::lock_guard<std::mutex> guard(mutex);
+    auto& queue = cuda_events[device];
+    if (!queue.empty()) {
+      *event = queue.front();
+      queue.pop();
+      return cudaSuccess;
+    }
+  }
+
+  return cudaEventCreateWithFlags(event, cudaEventBlockingSync |
+                                         cudaEventDisableTiming);
+}
+
+cudaError_t CUDAContext::ReleaseCudaEvent(cudaEvent_t event) {
+  int device;
+  auto status = cudaGetDevice(&device);
+  if (status != cudaSuccess) {
+    return status;
+  }
+
+  auto& mutex = cuda_events_mutex;
+  {
+    std::lock_guard<std::mutex> guard(mutex);
+    auto& queue = cuda_events[device];
+    queue.push(event);
+  }
+
+  return cudaSuccess;
+}
+
+void CUDAContext::ErrorCheck(std::string op_name, cudaError_t cuda_result) {
+  if (cuda_result != cudaSuccess) {
+    throw std::logic_error(std::string(op_name) + " failed: " + cudaGetErrorString(cuda_result));
+  }
+}
+
+void CUDAContext::RecordEvent(std::queue<std::pair<std::string, cudaEvent_t>>& event_queue,
+                              std::string name, cudaStream_t stream) {
+  cudaEvent_t event;
+  ErrorCheck("GetCudaEvent", GetCudaEvent(&event));
+  ErrorCheck("cudaEventRecord", cudaEventRecord(event, stream));
+  event_queue.emplace(name, event);
+}
+
+void CUDAContext::WaitForEvents(std::queue<std::pair<std::string, cudaEvent_t>>& event_queue,
+                                std::vector<TensorTableEntry>& entries, Timeline& timeline) {
+  while (!(event_queue).empty()) {
+    std::string name;
+    cudaEvent_t event;
+    std::tie(name, event) = (event_queue).front();
+    (event_queue).pop();
+    if (name != "") {
+      timeline.ActivityStartAll(entries, name);
+    }
+    ErrorCheck("cudaEventSynchronize", cudaEventSynchronize(event));
+    if (name != "") {
+      timeline.ActivityEndAll(entries);
+    }
+    ErrorCheck("ReleaseCudaEvent", ReleaseCudaEvent(event));
+  }
+}
+
+CUDAAllreduce::CUDAAllreduce(CUDAContext* context,
+                             CommunicationContext* comm_context,
+                             HorovodGlobalState* global_state)
+                             : AllreduceOp(comm_context, global_state), cuda_context_(context) {}
+
+Status CUDAAllreduce::Execute(std::vector<TensorTableEntry>& entries, const HorovodResponse& response) {
+  if (OnGPU(entries)) {
+    InitCUDA(entries);
+  }
+  return AllreduceOp::Execute(entries, response);
+}
+
+CUDACustomAllreduce::CUDACustomAllreduce(CUDAContext* context,
```

Comment:
AbstractCUDAAllreduce

Reply:
Changed to CUDAAllreduceAsync to better reflect the different behavior.